### PR TITLE
feat(Core/BGQueue): rework queue announce

### PR DESCRIPTION
**Pull request by Winfidonarleyan**
## Changes Proposed:
- Add manager for spam system
- Fix announce for arena skirmish
- Add config option `Arena.QueueAnnouncer.PlayerOnly`

## Issues Addressed:
- Part #5315

## Tests Performed:
-

## How to Test the Changes:
- Enable config options `Arena.QueueAnnouncer.Enable` and `Battleground.QueueAnnouncer.Enable`
- Try enter/exit rated arena
- Try enter/exit arena skirmish
- Try enter/exit bg (normal/random)
- Try https://github.com/azerothcore/mod-low-level-arena

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by FrancescoBorzi** [comment]
@Winfidonarleyan is this still WIP ? please let us know when it's ready to be tested ;)

**Comment by Winfidonarleyan** [comment]
It's all ready to go.

**Comment by FrancescoBorzi** [comment]
ok, I guess the test should make sure the Arena queue and announce systems still work fine.

@azerothcore/testers please have a look

**Comment by Winfidonarleyan** [comment]
I also think we should check the battleground

File: src/server/game/Battlegrounds/BattlegroundSpamProtect.cpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    bool IsCorrectDelay(ObjectGuid guid)
+    {
+        // Skip if spam time < 30 secs (default)
+        if (sWorld->GetGameTime() - GetTime(guid) < sWorld->getIntConfig(CONFIG_BATTLEGROUND_QUEUE_ANNOUNCER_SPAM_DELAY))
+        {
+            return false;
+        }
+
+        return true;
```
Comment (Helias) [inline src/server/game/Battlegrounds/BattlegroundSpamProtect.cpp:40]:
```suggestion
        // Skip if spam time < 30 secs (default)
        return sWorld->GetGameTime() - GetTime(guid) >= sWorld->getIntConfig(CONFIG_BATTLEGROUND_QUEUE_ANNOUNCER_SPAM_DELAY))
```

File: data/sql/updates/pending_db_world/rev_1622504741914151600.sql [hunk 2]
```diff
@@ -0,0 +1,6 @@
+INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1622504741914151600');
+
+DELETE FROM `acore_string` WHERE `entry` IN (713, 726);
+INSERT INTO `acore_string` (`entry`, `content_default`) VALUES
+(713, 'Queue status for %s (skirmish %s) (Lvl: %u to %u)\nQueued: %u (Need at least %u more)'),
+(726, '|cffff0000[Arena Queue]:|r %s (skirmish %s) -- [%u-%u] [%u/%u]|r');
```
Comment (FrancescoBorzi) [inline data/sql/updates/pending_db_world/rev_1622504741914151600.sql:6]: @locus313 @Kitzunu any idea about why this is failing the `check-pending-sql` script ?

Comment (Kitzunu) [reply data/sql/updates/pending_db_world/rev_1622504741914151600.sql:6]: Re running it.

**Comment by Sedatyf** [comment]
Is this still WIP or can I test it?

**Comment by Winfidonarleyan** [comment]
> Is this still WIP or can I test it?

Wip

**Comment by Winfidonarleyan** [comment]
> Is this still WIP or can I test it?

Can test this?

**Comment by Sedatyf** [comment]
Yeah sure, I'll take a look

**Comment by Sedatyf** [comment]
Hmmm... I think I'm missing something. I added ``mod-low-level-arena`` in my modules and then compiled it with cmake and visual. Cmake passed fine but Visual returned me like 117 errors. Do I have to add something else?

**Comment by Winfidonarleyan** [comment]
> Hmmm... I think I'm missing something. I added `mod-low-level-arena` in my modules and then compiled it with cmake and visual. Cmake passed fine but Visual returned me like 117 errors. Do I have to add something else?

To start, check without the module

**Comment by Sedatyf** [comment]
Alright, I tested it without the module. Everything looks fine! ✔️ 
I tagged in 2v2, 3v3, 5v5 in ranked and skirmish mode, I had an announce each time
I tagged in every battlegrounds as well and I had announcements too, except for Warsong Gulch. I'm not too sure if it's intended or not. Please see below screen for further informations
Screenshot 2021-06-10 164735
Screenshot 2021-06-10 165241

**Comment by Sedatyf** [comment]
Now, if you want me to test with the low level module, you may need to fix some stuff I guess. Or maybe explain to me what I'm missing because I had a shit tons of error when I compiled in VS

**Comment by Winfidonarleyan** [comment]
> Now, if you want me to test with the low level module, you may need to fix some stuff I guess. Or maybe explain to me what I'm missing because I had a shit tons of error when I compiled in VS

I'll look at the module later

**Comment by Winfidonarleyan** [comment]
Try to check the module now.
```
========== Сборка: успешно: 19, с ошибками: 0, без изменений: 0, пропущено: 1 ==========
```

**Comment by Sedatyf** [comment]
It seems to work for low level arena too!
image
I tried to do ranked arena with a team, but it said I had to be level 80 to have a team. So I don't know what was intended with low level arena modules, but announcements worked with low level arena skirmish!

File: src/server/game/Battlegrounds/BattlegroundQueue.cpp [hunk 3]
```diff
@@ -978,51 +986,97 @@ void BattlegroundQueue::SendMessageBGQueue(Player* leader, Battleground* bg, PvP
...
+    LOG_DEBUG("bg.battleground", "> Queue status for %s (Lvl: %u to %u) Queued: %u (Need at least %u more)",
+        bgName, q_min_level, q_max_level, qAlliance + qHorde, MaxPlayers - qTotal);
+
     // Show queue status to player only (when joining battleground queue or Arena and arena world announcer is disabled)
-    if (sWorld->getBoolConfig(CONFIG_BATTLEGROUND_QUEUE_ANNOUNCER_PLAYERONLY) || (bg->isArena() && !sWorld->getBoolConfig(CONFIG_ARENA_QUEUE_ANNOUNCER_ENABLE)))
+    if (sWorld->getBoolConfig(CONFIG_BATTLEGROUND_QUEUE_ANNOUNCER_PLAYERONLY))
     {
         ChatHandler(leader->GetSession()).PSendSysMessage(LANG_BG_QUEUE_ANNOUNCE_SELF, bgName, q_min_level, q_max_level,
-                qAlliance, (MinPlayers > qAlliance) ? MinPlayers - qAlliance : (uint32)0, qHorde, (MinPlayers > qHorde) ? MinPlayers - qHorde : (uint32)0);
+            qAlliance, (MinPlayers > qAlliance) ? MinPlayers - qAlliance : (uint32)0, qHorde, (MinPlayers > qHorde) ? MinPlayers - qHorde : (uint32)0);
```
Comment (Helias) [inline src/server/game/Battlegrounds/BattlegroundQueue.cpp:996]:
```suggestion
            qAlliance,
            (MinPlayers > qAlliance) ? MinPlayers - qAlliance : (uint32)0,
            qHorde, (MinPlayers > qHorde) ? MinPlayers - qHorde : (uint32)0
        );
```

**Review by Helias — APPROVED** [review]

**Review by FrancescoBorzi — APPROVED** [review]

**Pull request merged**